src/main/java/io/stargate/sgv2/jsonapi/service/embedding/VectorizeData.java

```diff
@@ -0,0 +1,114 @@
+package io.stargate.sgv2.jsonapi.service.embedding;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.stargate.sgv2.jsonapi.api.model.command.clause.sort.SortClause;
+import io.stargate.sgv2.jsonapi.api.model.command.clause.sort.SortExpression;
+import io.stargate.sgv2.jsonapi.api.model.command.clause.update.UpdateClause;
+import io.stargate.sgv2.jsonapi.api.model.command.clause.update.UpdateOperator;
+import io.stargate.sgv2.jsonapi.config.constants.DocumentConstants;
+import io.stargate.sgv2.jsonapi.exception.ErrorCode;
+import io.stargate.sgv2.jsonapi.exception.JsonApiException;
+import io.stargate.sgv2.jsonapi.service.embedding.operation.EmbeddingService;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class VectorizeData {
+  private final EmbeddingService embeddingService;
+  private final JsonNodeFactory nodeFactory;
+
+  public VectorizeData(EmbeddingService embeddingService, JsonNodeFactory nodeFactory) {
+    this.embeddingService = embeddingService;
+    this.nodeFactory = nodeFactory;
+  }
+
+  public void vectorize(List<JsonNode> documents) {
+    int vectorDataPosition = 0;
+    List<String> vectorizeTexts = new ArrayList<>();
+    Map<Integer, Integer> vectorizeMap = new HashMap<>();
+    for (int position = 0; position < documents.size(); position++) {
+      JsonNode document = documents.get(position);
+      if (document.has(DocumentConstants.Fields.VECTOR_EMBEDDING_TEXT_FIELD)) {
+        final JsonNode jsonNode =
+            document.get(DocumentConstants.Fields.VECTOR_EMBEDDING_TEXT_FIELD);
+        if (jsonNode.isNull()) {
+          ((ObjectNode) document)
+              .put(DocumentConstants.Fields.VECTOR_EMBEDDING_FIELD, (String) null);
+          continue;
+        }
+        if (!jsonNode.isTextual()) {
+          throw new JsonApiException(
+              ErrorCode.SHRED_BAD_VECTORIZE_VALUE,
+              ErrorCode.SHRED_BAD_VECTORIZE_VALUE.getMessage());
+        }
+
+        vectorizeTexts.add(jsonNode.asText());
+        vectorizeMap.put(vectorDataPosition, position);
+        vectorDataPosition++;
+      }
+    }
+
+    if (!vectorizeTexts.isEmpty()) {
+      List<float[]> vectors = embeddingService.vectorize(vectorizeTexts);
```

Comment:
Ah. So we collect vectorizable fields from all entries (within same collection), and make what is hopefully a batch call.

Reply:
Yes, in case of insert all, we collect all the $vectorize field from different document and call the embedding service as batch.

Reply:
That is cool. OpenAI/Vertex/Hugging Face, they all provide batch calls?